lib/mempool/mempool.c

```diff
@@ -0,0 +1,402 @@
+{
+	int i, from_l;
+	int alloc_l = -1, free_l = -1;
+	void *data;
+	size_t lsizes[p->n_levels];
+
+	/* Walk down through levels, finding the one from which we
+	 * want to allocate and the smallest one with a free entry
+	 * from which we can split an allocation if needed.  Along the
+	 * way, we populate an array of sizes for each level so we
+	 * don't need to waste RAM storing it.
+	 */
+	lsizes[0] = _ALIGN4(p->max_sz);
+	for (i = 0; i < p->n_levels; i++) {
+		if (i > 0) {
+			lsizes[i] = _ALIGN4(lsizes[i-1] / 4);
+		}
+
+		if (lsizes[i] < size) {
+			break;
+		}
+
+		alloc_l = i;
+		if (!level_empty(p, i)) {
+			free_l = i;
+		}
+	}
+
+	if (alloc_l < 0 || free_l < 0) {
+		*data_p = NULL;
+		return -ENOMEM;
+	}
+
+	/* Iteratively break the smallest enclosing block... */
+	data = block_alloc(p, free_l, lsizes[free_l]);
+
+	if (!data) {
+		/* This can happen if we race with another allocator.
+		 * It's OK, just back out and the timeout code will
+		 * retry.  Note mild overloading: -EAGAIN isn't for
+		 * propagation to the caller, it's to tell the loop in
+		 * k_mem_pool_alloc() to try again synchronously.  But
+		 * it means exactly what it says.
+		 *
+		 * This doesn't happen for user mode memory pools as this
+		 * entire function runs with a semaphore held.
+		 */
+		return -EAGAIN;
+	}
+
+	for (from_l = free_l; from_l < alloc_l; from_l++) {
+		data = block_break(p, data, from_l, lsizes);
+	}
+
+	*level_p = alloc_l;
+	*block_p = block_num(p, data, lsizes[alloc_l]);
+	*data_p = data;
+
+	return 0;
+}
+
+void _sys_mem_pool_block_free(struct sys_mem_pool_base *p, u32_t level,
+			      u32_t block)
+{
+	size_t lsizes[p->n_levels];
+	int i;
+
+	/* As in _sys_mem_pool_block_alloc(), we build a table of level sizes
+	 * to avoid having to store it in precious RAM bytes.
+	 * Overhead here is somewhat higher because block_free()
+	 * doesn't inherently need to traverse all the larger
+	 * sublevels.
+	 */
+	lsizes[0] = _ALIGN4(p->max_sz);
+	for (i = 1; i <= level; i++) {
+		lsizes[i] = _ALIGN4(lsizes[i-1] / 4);
+	}
+
+	block_free(p, level, lsizes, block);
+}
+
+
+/*
+ * Functions specific to user-mode blocks
+ */
+
+struct sys_mem_pool *sys_mem_pool_create(char *buf, size_t size,
+					 size_t minsz, size_t maxsz,
+					 struct k_mutex *mutex)
+{
+	size_t buf_size, lvls_size, bits_size;
+	struct sys_mem_pool *p;
+	struct sys_mem_pool_lvl *lvls;
+	u32_t n_max;
+
+	lvls_size = sizeof(struct sys_mem_pool_lvl) * _MPOOL_LVLS(maxsz, minsz);
+	buf_size = size - (sizeof(struct sys_mem_pool) + lvls_size);
+
+	p = (struct sys_mem_pool *)buf;
+	lvls = (struct sys_mem_pool_lvl *)(buf + sizeof(struct sys_mem_pool));
+	buf = buf + sizeof(struct sys_mem_pool) + lvls_size;
+
+	/* Iteratively compute space for both the actual memory blocks and
+	 * associated housekeeping structures, based on available remaining
+	 * size.
+	 *
+	 * XXX: A polynomial equation for _MPOOL_BITS_SIZE should make this
+	 * unnecessary.
```

Comment:
Not sure I love this API/algorithm.  If a user doesn't understand how maxsz interacts, they can easily end up with a broken heap where as much as half the space is wasted simply because they asked for a maxsz that was too big.  Might be better just to fail hard if the requested array doesn't fit, and make the user use a (cleaned up version of) the original size macros to allocate the static array.